docs/notebooks/explicit_constraints.pct.py

```diff
@@ -0,0 +1,190 @@
+
+# %%
+from trieste.models.gpflow import GaussianProcessRegression, build_gpr
+
+gpflow_model = build_gpr(initial_data, search_space, likelihood_variance=1e-7)
+model = GaussianProcessRegression(gpflow_model)
+
+
+# %% [markdown]
+# ## Acquisition function
+#
+# We can construct the _expected constrained improvement_ acquisition function similar to the [inequality-constraints notebook](inequality_constraints.ipynb). However, instead of using probability of feasibility with respect to the constraint model, we construct feasibility from the explicit input constraints. Feasibility is calculated by passing all the constraints residuals (to their respective limits) through a smoothing function and taking the product.
+#
+# Note this method penalises the expected improvement acquisition outside the feasible region. The optimizer uses unconstrained L-BFGS method to find the max of the acquistion function.
+
+# %%
+from trieste.acquisition.function import ExpectedConstrainedImprovement, FastConstraintsFeasibility
+from trieste.acquisition.rule import EfficientGlobalOptimization
+from trieste.observer import OBJECTIVE
+
+feas = FastConstraintsFeasibility(search_space)
+eci = ExpectedConstrainedImprovement(OBJECTIVE, feas.using(OBJECTIVE))
+
+rule = EfficientGlobalOptimization(eci)  # type: ignore
+
+# %% [markdown]
+# ## Run the optimization loop
+#
+# We can now run the optimization loop.
+
+# %%
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+
+result = bo.optimize(15, initial_data, model, acquisition_rule=rule)
+
+# %% [markdown]
+# We can now get the best point found by the optimizer. Note this isn’t necessarily the point that was last evaluated.
+
+# %%
+query_point, observation, arg_min_idx = result.try_get_optimal_point()
+
+print(f"query point: {query_point}")
+print(f"observation: {observation}")
+
+# %% [markdown]
+# We obtain the final objective and constraint data using `.try_get_final_datasets()`. We can visualise how the optimizer performed by plotting all the acquired observations, along with the true function values and optima.
+#
+# The crosses are the 5 initial points that were sampled from the entire search space. The green circles are the acquired observations by the optimizer. The purple circle is the best point found.
+
+# %%
+from trieste.experimental.plotting import plot_bo_points, plot_function_2d
+
+dataset = result.try_get_final_dataset()
+query_points = dataset.query_points.numpy()
+observations = dataset.observations.numpy()
+
+_, ax = plot_function_2d(
+    ScaledBranin.objective,
+    search_space.lower,
+    search_space.upper,
+    grid_density=30,
+    contour=True,
+)
+plot_bo_points(query_points, ax[0, 0], num_initial_points, arg_min_idx)
```

Comment:
The plot is a bit unclear for me: maybe make the dots a bit bigger and the purple one brighter? (Alternatively, ask @vpicheny for a second opinion!)

Reply:
That is an existing plotting function. I don't think the point size can be changed from its args but the colour can be, which I now have. Also, I have made the overall figure larger. Hopefully that is good enough.